[OpenMP] Fix td_tdg_task_id underflow when taskloop and taskgraph

jpinot opened the pull request:
This patch addresses an issue where the `td_tdg_task_id` could underflow, leading to a negative task ID, when a `taskloop` region was encountered before a taskgraph clause.

Previously, `td_tdg_task_id` was decremented unconditionally within a taskloop, even if that taskloop was not part of an active taskgraph. This caused problems in scenarios like the following:

```cpp
#pragma omp parallel
#pragma omp single
{
// Here, td_tdg_task_id is incorrectly decremented, potentially becoming negative.
#pragma omp taskloop
  for (int i = 0; i < n; i++) {
    initialize(data[i]);
  }
}

#pragma omp parallel
#pragma omp single
{
#pragma omp taskgraph
{
// When this taskgraph is entered, the recording of tasks may fail (e.g., segfault)
// due to the invalid (negative) td_tdg_task_id.
#pragma omp taskloop
  for (int i = 0; i < num_iter;  i++) {
    compute();
  }
}
}
```

**EDIT(18/09/2025)**:
This implementation:
- Allows sporadic holes in the record_map: The record_map, an array of recorded tasks in a task graph, can now have sporadic holes. This change removes the need for a decrement operation and centralizes the control of the index/ID within each task dependence graph (TDG) by moving the count for the next index inside the kmp_node_info structure.
- Removes extra allocation of kmp_node_info successors. Optimizing memory usage by delaying the allocation of kmp_node_info successors until they are actually needed.
- Fixes a data race when resizing the record_map.

jpinot committed:
[OpenMP] Fix td_tdg_task_id underflow with taskloop and taskgraph

This patch addresses an issue where the td_tdg_task_id could underflow,
leading to a negative task ID, when a taskloop region was encountered
before a taskgraph clause.

This change allows surious holes in the record_map.

jprotze commented:
Even with a localized counter as implemented with your latest commit, I think this decrement can be problematic.

Is it really important to avoid spurious holes in the numbering of task IDs? As I understand, this is just used for accessing the `record_map` entries (which is actually a vector, not a map). Checking the code, I realized that for resizing the vector, the code will do a lot of rather small allocs for `successorList`, that might never be used. I would recommend to delay these allocs until the entry is really used.

A bigger issue is, that the accesses to the `record_map` are not protected. If one thread triggers the resize code for the vector, other threads might still access the vector and writes might get lost.

jpinot committed:
[openmp] Remove taskgraph successors alloction before is needed

Delayed allocation of successors in kmp_node until the array is needed,
removing the small allocation when a taskgraph node is created or
resized.

jpinot committed:
[openmp] Fix perent_task unitialization when allocating kmp_tdg_info

jpinot committed:
[openmp] Fix locking when expanding recorded tdg

jpinot commented:
> I think, the number of places where you iterate over all records is limited. In these places, you could simply check for `record_map[i].task != NULL` and skip the spurious holes in the vector of records. In other places where the record map is used to find successors, you can still blindly access the map entries.

Thank you for the feedback and sorry for the delayed response.
I agree that allowing spurious holes in the `record_map` simplifies the logic. I’ve updated the branch to incorporate your suggestion. I’ve also added the extra locking mechanism for resizing `record_map` and `successors`, and removed the premature allocation of `successors`. My preliminary tests have been successful.

@jprotze, what are your thoughts?

jpinot committed:
[openmp] Add tdg node pointer in kmp_taskdata_t

Add pointer to node represeting the task in the TDG, the way avoids
locking access to record_map every time a node need to be accessed.

jpinot commented:
> The `record_map` accesses in `__kmp_track_dependence` are not protected for a concurrent reallocation of the vector. I didn't check all the other accesses to the `record_map`, whether they might be concurrent to a reallocation.
> 
> I think there are two options:
> 
>     * lock any access to `record_map` that is possibly concurrent to a write/push (probably way too expensive)
> 
>     * store a pointer to the node_info in kmp_taskdata_t: `task->tdg_node_info = task->tdg->record_map[task->td_tdg_task_id]`

Yes, you're right. I agree that locking all read/write interactions with record_map would be too expensive.
The problem with your second suggestion (storing a pointer to the node in taskdata) is that record_map is an array of kmp_node_info_t. When the record_map vector is reallocated(resize), the old address is freed, and the pointer would point to a deleted memory location.

Two possible solutions come to mind:
 - Change the vector to a linked list. This would solve the pointer invalidation issue but could decrease efficiency during linear traversals, e.g: when executing a recorded TDG.
 - Use a multi-dimensional array or a similar data structure that avoids frequent reallocation, thus preventing the addresses from changing.
 
 @jprotze any thoughts?

jpinot committed:
[OpenMP] Add kmp_node_vector for TDG successors

Replaced the fixed-size array for TDG successors with kmp_node_vector,
a custom dynamic vector of kmp_node_info to TDG nodes. This change aims
to mitigate data races during vector resizing by using a block-based
allocation strategy.

jpinot commented:
> Right, I missed, that record_map is essentially of type `vector<kmp_node_info_t>`, not `vector<kmp_node_info_t *>`.
> 
> Another possible approach would be to use a customized vector which allocates `fixed_sized` blocks as necessary (instead of realloc) and redirects `operator[](i)` to look at the `i % fixed_sized` element of block `i / fixed_sized`.

I have pushed a WIP commit with the changes discussed. However, a data race occurs when thread A reallocates blocks (the array of `blocks`; note that the actual blocks are never resized) to increase the number of blocks, and thread B calls  [kmp_node_vector_get](https://github.com/llvm/llvm-project/pull/150877/commits/b151a84f336838f5bdc0f41b1c2798fd2f3fe22e#diff-8402e656316eb873d5db4dea7f697406d15ae4197dcc60d88b3d9fc252fcb69aR5377) to access an existing node. The returned pointer points to invalid data. I suspect this could be due to a compiler optimization or an issue with the return value handling(?).

@jprotze, do you know what might be causing this?

jpinot committed:
[openmp] Delete graph_lock

Munesanz commented:
Hi @jprotze, what are your thoughts on the last jpinot changes? thanks
